src/devices/SoftPwm/SoftwarePwmChannel.cs

```diff
@@ -105,106 +110,89 @@ public SoftwarePwmChannel(int pinNumber, int frequency = 400, double dutyCycle =
-                    if (_pulseWidthInMilliseconds != 0)
-                    {
-                        _controller.Write(_pin, PinValue.High);
-                        _isStopped = false;
-                    }
-
-                    // Use the wait helper method to wait for the length of the pulse
-                    if (_usePrecisionTimer)
-                    {
-                        Wait(_pulseWidthInMilliseconds);
-                    }
-                    else
-                    {
-                        Task.Delay(TimeSpan.FromMilliseconds(_pulseWidthInMilliseconds)).Wait();
-                    }
-
-                    // The pulse if over and so set the pin to low and then wait until it's time for the next pulse
-                    _controller.Write(_pin, PinValue.Low);
+                    Thread.Yield();
+                    continue;
+                }
 
-                    if (_usePrecisionTimer)
-                    {
-                        Wait(_periodInMilliseconds - _pulseWidthInMilliseconds);
-                    }
-                    else
-                    {
-                        Task.Delay(TimeSpan.FromMilliseconds(_periodInMilliseconds - _pulseWidthInMilliseconds)).Wait();
-                    }
+                if (_pinHighTime != TimeSpan.Zero)
+                {
+                    _controller.Write(_pin, PinValue.High);
+                    Wait(_pinHighTime);
                 }
-                else
+
+                if (_pinLowTime != TimeSpan.Zero)
                 {
-                    if (!_isStopped)
-                    {
-                        _controller.Write(_pin, PinValue.Low);
-                        _isStopped = true;
-                    }
+                    _controller.Write(_pin, PinValue.Low);
+                    Wait(_pinLowTime);
                 }
             }
+
+            _controller.Write(_pin, PinValue.Low);
         }
 
         /// <summary>
         /// A synchronous wait is used to avoid yielding the thread
         /// This method calculates the number of CPU ticks will elapse in the specified time and spins
         /// in a loop until that threshold is hit. This allows for very precise timing.
         /// </summary>
-        /// <param name="milliseconds">The milliseconds to wait for</param>
-        private void Wait(double milliseconds)
+        /// <param name="ticks">The time to wait.</param>
+        private void Wait(TimeSpan timeout)
         {
-            long initialTick = _stopwatch.ElapsedTicks;
-            long initialElapsed = _stopwatch.ElapsedMilliseconds;
-            double desiredTicks = milliseconds / 1000.0 * Stopwatch.Frequency;
-            double finalTick = initialTick + desiredTicks;
-            while (_stopwatch.ElapsedTicks < finalTick)
+            if (_stopwatch is null)
+            {
+                Thread.Sleep(timeout);
+                return;
+            }
+
+            TimeSpan elapsed = _stopwatch.Elapsed;
```

Comment:
The old logic was trying to use the raw ticks, but wasn't  necessarily account for large or small frequencies, which is where the logic was broken.
Some devices, for example, report a 1ns frequency while others report 100ns and others may report even lower resolutions (such as 1 micro or 1 millisecond).
The new logic just relies on the existing `Stopwatch.Elapsed` logic which normalizes the "raw ticks" into "timespan ticks" so everything can be cleanly and correctly handled.